include/alpaka/mem/buf/BufCpu.hpp

```diff
@@ -81,24 +79,12 @@ namespace alpaka
                 // Unpin this memory if it is currently pinned.
                 unpin(*this);
 #endif
-                // NOTE: m_pMem is allowed to be a nullptr here.
-                alpaka::free(*this, m_pMem);
-            }
-
-        private:
-            //! \return The number of elements to allocate.
-            template<typename TExtent>
-            ALPAKA_FN_HOST static auto computeElementCount(TExtent const& extent) -> TIdx
-            {
-                auto const extentElementCount = extent::getExtentProduct(extent);
-
-                return extentElementCount;
             }
 
         public:
             DevCpu const m_dev;
             Vec<TDim, TIdx> const m_extentElements;
-            TElem* const m_pMem;
+            std::shared_ptr<TElem> const m_spMem;
```

Comment:
This approach does add a small overhead, replacing a `TElem*` with a `std::shared_ptr<TElem>`.
This could be avoided by storing the deleter directly in an `std::function` instead of using the `shared_ptr` for that.

An other possibility would be to follow the same approach used by `BufUniformCudaHipRt`:
  - move the metadata (device, extent, pitch) from `BufCpuImpl` directly into `BufCpu`
  - replace the `shared_ptr<BufCpuImpl>` with a `shared_ptr<TElem>`, and set the destructor there

However, this would trade the one less indirection with more expensive copies for the `BufCpu` objects, since the metadata would have to be copied as well.

Reply:
Storing the deleter explicitly proved to be simple enough, and should add a smaller overhead than using an extra `shared_ptr`.